### PR TITLE
Add resource  annotations for parameterizedTemplate

### PR DESCRIPTION
**Pull request by fisherxu**
Currently we can't set the resource for sidecar in GenerateTemplateFromParams. So add resource annotations to the parameterizedTemplate, also add the UT for it.

### CODE REVIEW

**Comment by nmittler** [comment]
You should also update the registry: https://github.com/istio/istio/blob/master/pilot/pkg/kube/inject/inject.go#L53

It's not strictly necessary, but it's a catalog of the annotations that we support.

**Comment by fisherxu** [comment]
@nmittler  Have updated, PTAL :)

File: pilot/pkg/kube/inject/mesh.go [hunk 1]
```diff
@@ -156,8 +158,14 @@ containers:
   imagePullPolicy: {{ .ImagePullPolicy }}
   {{ end -}}
   resources:
+    {{ "[[ if (isset .ObjectMeta.Annotations $proxyCPUKey) -]]" }}
```
Comment (nmittler) [inline pilot/pkg/kube/inject/mesh.go]: Create a variable at the top of the yaml to hold the value of the annotation or it's default:

```yaml
 [[- $proxyCPUValue := (annotation .ObjectMeta $proxyCPUKey "10m") -]]
 [[- $proxyMemoryValue := (annotation .ObjectMeta $proxyMemoryKey "") -]]
```
Then the code here becomes something like:

```yaml
requests:
  cpu:  {{ "\"[[ $proxyCPUValue ]]\"" }}
  {{ "[[ if (ne $proxyMemoryValue '') -]]" }}
  memory: {{ "\"[[ $proxyMemoryValue ]]\"" }}
  {{ "[[- end ]]" }}
```

File: pilot/pkg/kube/inject/testdata/inject/resource_annotations.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+apiVersion: extensions/v1beta1
```
Comment (nmittler) [inline pilot/pkg/kube/inject/testdata/inject/resource_annotations.yaml]: just to confirm: the new files under `/testdata/inject` are identical to the ones under `/testdata/webhook`?

**Comment by fisherxu** [comment]
@nmittler https://github.com/istio/istio/blob/master/install/kubernetes/helm/istio/templates/sidecar-injector-configmap.yaml#L154 don't understand update what? :)

**Comment by nmittler** [comment]
@fisherxu 

> @nmittler https://github.com/istio/istio/blob/master/install/kubernetes/helm/istio/templates/sidecar-injector-configmap.yaml#L154 don't understand update what? :)

Oh, had you already submitted that change in a previous PR? It looks like that logic is wrong.  The template as you have it assumes that the annotation exists.  You need to change this to the logic that I proposed in mesh.go.

**Comment by nmittler** [comment]
@fisherxu never mind ... looks like these annotations already existed.  I think we need to fix these lines, however.  We need to use appropriate defaults if the annotations aren't present. Like what we're doing in mesh.go.

**Comment by fisherxu** [comment]
> We need to use appropriate defaults if the annotations aren't present. Like what we're doing in mesh.go.

**Comment by fisherxu** [comment]
> We need to use appropriate defaults if the annotations aren't present. Like what we're doing in mesh.go.

@nmittler  In sidecar-injector-configmap.yaml, I think the logic is when user set the annotations, then will read the value. And if not set it, will use the default value from helm global value.
so not sure how to fix it? :)

File: pilot/pkg/kube/inject/mesh.go (hunk 1 above)
Comment (fisherxu) [reply pilot/pkg/kube/inject/mesh.go]: Done.

File: pilot/pkg/kube/inject/testdata/inject/resource_annotations.yaml (hunk 2 above)
Comment (fisherxu) [reply pilot/pkg/kube/inject/testdata/inject/resource_annotations.yaml]: Yeah, it's same with the one under `/testdata/webhook`, but some of other files are also same. Do we need to fix all of them?

**Comment by nmittler** [comment]
@fisherxu for helm, the defaults are defined in [values.yaml](https://github.com/istio/istio/blob/master/install/kubernetes/helm/istio/values.yaml#L24).  For example, you'll see that the `statusPort` is referenced from [sidecar-injector-configmap.yaml](https://github.com/istio/istio/blob/master/install/kubernetes/helm/istio/templates/sidecar-injector-configmap.yaml#L132) as `Values.global.proxy.statusPort`.

You can look throughout the file for other examples where `Values` properties are used as defaults for the `annotation` function.

**Comment by fisherxu** [comment]
@nmittler I have fix it in the `sidecar-injector-configmap.yaml`, and the test case have passed, thanks for your suggestion, PTAL.

File: install/kubernetes/helm/istio-remote/templates/sidecar-injector-configmap.yaml [hunk 3]
```diff
@@ -146,15 +146,11 @@ data:
...
           requests:
-            cpu: {{ "\"[[ index .ObjectMeta.Annotations \"sidecar.istio.io/proxyCPU\" ]]\"" }}
-            memory: {{ "\"[[ index .ObjectMeta.Annotations \"sidecar.istio.io/proxyMemory\" ]]\"" }}
-        {{ "[[ else -]]" }}
-{{- if .Values.global.proxy.resources }}
-{{ toYaml .Values.global.proxy.resources | indent 10 }}
-{{- end }}
-        {{ "[[ end -]]" }}
+            cpu: {{ "[[ annotation .ObjectMeta `sidecar.istio.io/proxyCPU` " }} "{{ .Values.global.proxy.resources.requests.cpu }}" {{ " ]]" }}
+            {{ "[[ if ne (annotation .ObjectMeta `sidecar.istio.io/proxyMemory` \"\") \"\" -]]" }}
```
Comment (nmittler) [inline install/kubernetes/helm/istio-remote/templates/sidecar-injector-configmap.yaml]: do we want to establish a default value for memory in the configmap as well?

@ostromart thoughts?

File: pilot/pkg/kube/inject/inject_test.go [hunk 4]
```diff
@@ -473,6 +473,17 @@ func TestIntoResourceFile(t *testing.T) {
 			readinessPeriodSeconds:       DefaultReadinessPeriodSeconds,
 			readinessFailureThreshold:    DefaultReadinessFailureThreshold,
 		},
+		{
```
Comment (nmittler) [inline pilot/pkg/kube/inject/inject_test.go]: I think the test for webhook was already merged?

File: install/kubernetes/helm/istio-remote/templates/sidecar-injector-configmap.yaml (hunk 3 above)
Comment (ostromart) [reply install/kubernetes/helm/istio-remote/templates/sidecar-injector-configmap.yaml]: Yes, I also think 10m might be too low (below). @costinm has been doing scale testing so he probably has the best idea for good default values here.

File: pilot/pkg/kube/inject/mesh.go [hunk 5]
```diff
@@ -36,6 +38,8 @@ const (
 [[- $readinessPeriodValue           := (annotation .ObjectMeta $readinessPeriodKey "{{ .ReadinessPeriodSeconds }}") ]]
 [[- $readinessFailureThresholdValue := (annotation .ObjectMeta $readinessFailureThresholdKey {{ .ReadinessFailureThreshold }}) -]]
 [[- $readinessApplicationPortsValue := (annotation .ObjectMeta $readinessApplicationPortsKey (applicationPorts .Spec.Containers)) -]]
+[[- $proxyCPUValue                  := (annotation .ObjectMeta $proxyCPUKey "10m") -]]
```
Comment (ostromart) [inline pilot/pkg/kube/inject/mesh.go]: Too low?

File: pilot/pkg/kube/inject/inject_test.go (hunk 4 above)
Comment (fisherxu) [reply pilot/pkg/kube/inject/inject_test.go]: Yes, I have pick it here,  ready to close it but an owner has merged it yesterday...

File: pilot/pkg/kube/inject/mesh.go (hunk 5 above)
Comment (fisherxu) [reply pilot/pkg/kube/inject/mesh.go]: +1, 10m maybe to low...

File: install/kubernetes/helm/istio-remote/templates/sidecar-injector-configmap.yaml (hunk 3 above)
Comment (fisherxu) [reply install/kubernetes/helm/istio-remote/templates/sidecar-injector-configmap.yaml]: I think we should also need to implement the ability to config the resource limit? It's related to the pod QoS class in k8s.

**Comment by nmittler** [comment]
@ayj @ostromart you guys want to take a look as well?

File: install/kubernetes/helm/istio-remote/templates/sidecar-injector-configmap.yaml [hunk 6]
```diff
@@ -146,15 +146,11 @@ data:
...
-          {{ "[[ if (isset .ObjectMeta.Annotations \"sidecar.istio.io/proxyCPU\") -]]" }}
           requests:
-            cpu: {{ "\"[[ index .ObjectMeta.Annotations \"sidecar.istio.io/proxyCPU\" ]]\"" }}
-            memory: {{ "\"[[ index .ObjectMeta.Annotations \"sidecar.istio.io/proxyMemory\" ]]\"" }}
-        {{ "[[ else -]]" }}
-{{- if .Values.global.proxy.resources }}
-{{ toYaml .Values.global.proxy.resources | indent 10 }}
-{{- end }}
-        {{ "[[ end -]]" }}
+            cpu: {{ "[[ annotation .ObjectMeta `sidecar.istio.io/proxyCPU` " }} "{{ .Values.global.proxy.resources.requests.cpu }}" {{ " ]]" }}
```
Comment (sdake) [inline install/kubernetes/helm/istio-remote/templates/sidecar-injector-configmap.yaml]: @ayj can you have a review of the install section?

**Comment by catherinetcai** [comment]
Would it be possible to revive this? Running into this on Istio 1.0.2.

**Comment by ostromart** [comment]
@fisherxu any interest in reopening it? I can take a look if there are no volunteers.

**Comment by ostromart** [comment]
See https://github.com/istio/istio/issues/9332#issuecomment-461099325 and https://github.com/istio/istio/issues/8553 and

**Pull request closed without merging**